File: POO/Constructores_Clases/Abstract class/js/Notificacion.js
class Notificacion {
    constructor() {
        if (this.constructor === Notificacion) {
            throw new Error("No se puede instanciar una clase abstracta");
        }
    }

    enviar() {
        throw new Error("Método abstracto debe ser implementado");
    }

    obtenerPrioridad() {
        throw new Error("Método abstracto debe ser implementado");
    }

    mostrar() {
        throw new Error("Método abstracto debe ser implementado");
    }
}

class SMS extends Notificacion {
    constructor(numero, mensaje, urgente) {
        super();
        this.numero = numero;
        this.mensaje = mensaje;
        this.urgente = urgente;
    }

    enviar() {
        return `SMS enviado al número: ${this.numero}`;
    }

    obtenerPrioridad() {
        return this.urgente ? "ALTA" : "NORMAL";
    }

    mostrar() {
        console.log(`SMS para: ${this.numero}`);
        console.log(`Mensaje: ${this.mensaje}`);
        console.log(`Prioridad: ${this.obtenerPrioridad()}`);
    }
}

// Uso
const sms = new SMS("+123456789", "Tu código de verificación es: 1234", true);
sms.mostrar();
console.log(sms.enviar());

File: POO/Constructores_Clases/Abstract class/GO/Notificacion.go
package main

import "fmt"

type Notificacion interface {
    Enviar() string
    Prioridad() string
    Mostrar()
}

type Email struct {
    destinatario string
    asunto       string
    mensaje      string
    urgente      bool
}

// Constructor con parámetros
func NewEmail(destinatario, asunto, mensaje string, urgente bool) *Email {
    return &Email{
        destinatario: destinatario,
        asunto:       asunto,
        mensaje:      mensaje,
        urgente:      urgente,
    }
}

func (e *Email) Enviar() string {
    return fmt.Sprintf("Email enviado a: %s", e.destinatario)
}

func (e *Email) Prioridad() string {
    if e.urgente {
        return "ALTA"
    }
    return "NORMAL"
}

func (e *Email) Mostrar() {
    fmt.Printf("Email para: %s\n", e.destinatario)
    fmt.Printf("Asunto: %s, Prioridad: %s\n", e.asunto, e.Prioridad())
    fmt.Printf("Mensaje: %s\n", e.mensaje)
}

func main() {
    var notif Notificacion = NewEmail(
        "cliente@empresa.com", 
        "Importante: Reunión", 
        "Tenemos una reunión mañana a las 10:00 AM", 
        true,
    )
    notif.Mostrar()
}